evalml/tests/component_tests/test_target_imputer.py

```diff
@@ -0,0 +1,185 @@
+import numpy as np
+import pandas as pd
+import pytest
+import woodwork as ww
+from pandas.testing import assert_series_equal
+from woodwork.logical_types import (
+    Boolean,
+    Categorical,
+    Double,
+    Integer,
+    NaturalLanguage
+)
+
+from evalml.pipelines.components import TargetImputer
+
+
+def test_target_imputer_no_y(X_y_binary):
+    X, y = X_y_binary
+    imputer = TargetImputer(impute_strategy='median')
+    with pytest.raises(ValueError, match="y cannot be None"):
+        imputer.fit_transform(None, None)
+    with pytest.raises(ValueError, match="y cannot be None"):
+        imputer.fit(X, None)
+
+
+def test_target_imputer_median():
+    y = pd.Series([np.nan, 1, 10, 10, 6])
+    imputer = TargetImputer(impute_strategy='median')
+    y_expected = pd.Series([8, 1, 10, 10, 6])
+    _, y_t = imputer.fit_transform(None, y)
+    assert_series_equal(y_expected, y_t.to_series(), check_dtype=False)
+
+
+def test_target_imputer_mean():
+    y = pd.Series([np.nan, 2, 0])
+    imputer = TargetImputer(impute_strategy='mean')
+    y_expected = pd.Series([1, 2, 0])
+    _, y_t = imputer.fit_transform(None, y)
+    assert_series_equal(y_expected, y_t.to_series(), check_dtype=False)
+
+
+@pytest.mark.parametrize("fill_value, y, y_expected", [(None, pd.Series([np.nan, 0, 5]), pd.Series([0, 0, 5])),
+                                                       (None, pd.Series([np.nan, "a", "b"]), pd.Series(["missing_value", "a", "b"]).astype("category")),
+                                                       (3, pd.Series([np.nan, 0, 5]), pd.Series([3, 0, 5])),
+                                                       (3, pd.Series([np.nan, "a", "b"]), pd.Series([3, "a", "b"]).astype("category"))])
```

Comment:
This last parametrized test case is a very interesting one.  Do we want to match types?  Like if the integer 3 is put in, do we want it filling with the integer 3?  Or the string 3?  Do we want to allow cross-type imputation?  Or perhaps raise a value error.

Reply:
Yeah, that's a good question! This follows the behavior in SimpleImputer / Imputer right now, but I think it's okay because the type of the series is `category`, which allows for mixed-type categories:

![image](https://user-images.githubusercontent.com/5422235/112893833-2d5eea80-90a9-11eb-8e5e-27b37126953e.png)

Happy to file an issue if you think this is worth a greater discussion though!